### PR TITLE
crimson/asock: cleanups

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/admin/admin_socket.cc [hunk 1]
```diff
@@ -44,25 +44,14 @@ tell_result_t::tell_result_t(std::unique_ptr<Formatter> formatter)
   formatter->flush(out);
 }
 
-seastar::future<>
-AdminSocket::register_command(std::unique_ptr<AdminSocketHook>&& hook)
+void AdminSocket::register_command(std::unique_ptr<AdminSocketHook>&& hook)
 {
-  return seastar::with_lock(servers_tbl_rwlock,
```
Comment (rzarzynski) [inline src/crimson/admin/admin_socket.cc:50]: This simplification looks sane. I would just write the requirement (no late registration) down.

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/admin/admin_socket.cc [hunk 2]
```diff
@@ -371,20 +355,18 @@ class GetdescsHook final : public AdminSocketHook {
...
+    for (const auto& [prefix, hook] : m_as) {
+      auto secname = fmt::format("cmd {:>03}", cmdnum);
+      auto cmd = fmt::format("{} {}", hook->prefix, hook->desc);
+      dump_cmd_and_help_to_json(f.get(), CEPH_FEATURES_ALL, secname,
+				cmd, std::string{hook->help});
+      cmdnum++;
+    }
+    f->close_section();
+    return seastar::make_ready_future<tell_result_t>(std::move(f));
   }
```
Comment (liu-chunmei) [inline src/crimson/admin/admin_socket.cc:370]: can we make this not return future?

File: src/crimson/admin/admin_socket.cc [hunk 3]
```diff
@@ -342,18 +328,16 @@ class HelpHook final : public AdminSocketHook {
...
+      if (!hook->help.empty()) {
+        f->dump_string(prefix.data(), hook->help);
       }
-      f->close_section();
-      return seastar::make_ready_future<tell_result_t>(std::move(f));
-    });
+    }
+    f->close_section();
+    return seastar::make_ready_future<tell_result_t>(std::move(f));
   }
```
Comment (liu-chunmei) [inline src/crimson/admin/admin_socket.cc:341]: can we make this not return future?

File: src/crimson/admin/admin_socket.cc [hunk 4]
```diff
@@ -155,26 +144,23 @@ auto AdminSocket::execute_command(const std::vector<std::string>& cmd,
...
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, "invalid command json", std::move(out)});
     }
-  });
+    return parsed->hook.call(parsed->params, parsed->format, std::move(buf));
+  } else {
+    auto& result = std::get<tell_result_t>(maybe_parsed);
+    return seastar::make_ready_future<tell_result_t>(std::move(result));
+  }
 }
```
Comment (liu-chunmei) [inline src/crimson/admin/admin_socket.cc:164]: same here. not return future?

Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc:164]: no, we cannot assume that `hook.call()` is always synchronous .

File: src/crimson/admin/admin_socket.cc (hunk 3 above)
Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc:341]: please note, `call()` is a virtual function. we could have an extra layer allowing an async `call()` calls into the synchronous version, and convert the plain return value into a `seastar::future<>`. but i want to make this simple.

File: src/crimson/admin/admin_socket.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc:370]: ditto.

File: src/crimson/admin/admin_socket.cc (hunk 1 above)
Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc:50]: i think actually it's fine to register hooks after the admin sock infra starts. as all the access to the `hooks` are synchronous. and the `hook` is valid after the `hooks` is mutated.

**Pull request merged**